pkg/reconciler/subscription/subscription.go

```diff
@@ -512,3 +527,37 @@ func deliverySpec(sub *v1.Subscription, channel *eventingduckv1.Channelable) (de
 	}
 	return
 }
+
+// TODO(slinkydeveloper) this can be refactored in a separate object that takes care of "ref resolution"
+func (r *Reconciler) resolveRefAPIVersion(objRef *corev1.ObjectReference) error {
+	if objRef.APIVersion == "" || strings.Contains(objRef.APIVersion, "/") {
+		// Either it's Core v1 or the version is specified manually
+		return nil
+	}
+
+	actualGvk := schema.GroupVersionKind{Group: objRef.APIVersion, Kind: objRef.Kind}
+	pluralGvk, _ := meta.UnsafeGuessKindToResource(actualGvk)
+	crd, err := r.crdLister.Get(pluralGvk.GroupResource().String())
+	if err != nil {
+		return err
+	}
+
+	actualGvk.Version, err = findCRDStorageVersion(crd)
```

Comment:
I am not sure this will do what you want. What you are trying to do is find the version of the channel that implements the duck type the subscription controller assumes to use, but this is not the way to do it. The stored version means nothing to the duck type.

Reply:
> What you are trying to do is find the version of the channel that implements the duck type the subscription controller assumes to use, but this is not the way to do it

What is the proper way, if there is any?

> The stored version means nothing to the duck type.

From my understanding, the duck type doesn't care neither of the stored version, nor the original version used by the user to create the resource.

To be more specific: here I'm not performing discovery of duck types, nor looking for the duck type version etc. The user gives me a concrete channel kind and api group and I select for her/him the version, that's it. For what is worth, we could even pick just the first or the last version, since which version to pick is not important.

I query the storage version because, at the end of the day, this is the ultimate source of truth for k8s regarding that particular resource and skips some conversions in the middle.

Reply:
Proper way is to specify the API version that you want to use. The label on the resource says which duck type if conforms to. Storage version is totally irrelevant, just in this case you happen to know that this version conforms to a duck type.

Reply:
Idea in this PR:
```
Kind: Subscription
...
Spec:
  Channel:
    API Version:  messaging.knative.dev
    Kind:         KafkaChannel
    Name:         testchannel
```

When the controller sees that there's no version in the "API version" field, it goes and checks the versions of this resource.  It **transiently** sets the storage version of the resource on the reference. So, in memory, it becomes like this:
```
Kind: Subscription
...
Spec:
  Channel:
    API Version:  messaging.knative.dev/v1alpha1
    Kind:         KafkaChannel
    Name:         testchannel
```

And then, it fetches the channel as `messaging.knative.dev/v1alpha1` KafkaChannel.

In KafkaChannel, we have `v1alpha1` API supporting `v1alpha1` subscribable duck and we have `v1beta1` API supporting `v1` subscribable duck.

When it fetches the channel as `messaging.knative.dev/v1alpha1`, doesn't it get `v1alpha1` in the duck annotation? And then based on that the subscriber reconciler would operate as if this channel ref is a ref to a channel that's supporting `v1alpha1` duck type.

This is why the storage version is used. If there's no version is specified, it just falls back to storage version to be able to operate with that resource.

> What you are trying to do is find the version of the channel that implements the duck type the subscription controller assumes to use, but this is not the way to do it.

Nope, just find an arbitrary version (in this case, getting the storage version makes the reasonable arbitrary version to use) and then let the subscription controller take it from there. It can handle whatever duck version is specified in the resource fetched as storage version.